runtime/parachains/src/paras.rs

```diff
@@ -534,9 +676,175 @@ pub mod pallet {
+			}
+
+			Ok(())
+		}
+	}
+
+	#[pallet::validate_unsigned]
+	impl<T: Config> ValidateUnsigned for Pallet<T> {
+		type Call = Call<T>;
+
+		fn validate_unsigned(_source: TransactionSource, call: &Self::Call) -> TransactionValidity {
+			let (stmt, signature) = match call {
+				Call::include_pvf_check_statement { stmt, signature } => (stmt, signature),
+				_ => return InvalidTransaction::Call.into(),
+			};
+
+			let current_session = shared::Pallet::<T>::session_index();
+			if stmt.session_index < current_session {
+				return InvalidTransaction::Stale.into()
+			} else if stmt.session_index > current_session {
+				return InvalidTransaction::Future.into()
+			}
+
+			let validator_index = stmt.validator_index.0 as usize;
+			let validators = shared::Pallet::<T>::active_validator_keys();
+			let validator_public = match validators.get(validator_index) {
+				Some(pk) => pk,
+				None => return InvalidTransaction::Custom(INVALID_TX_BAD_VALIDATOR_IDX).into(),
+			};
+
+			let signing_payload = stmt.signing_payload();
+			if !signature.verify(&signing_payload[..], &validator_public) {
+				return InvalidTransaction::BadProof.into()
+			}
+
+			let active_vote = match PvfActiveVoteMap::<T>::get(&stmt.subject) {
+				Some(v) => v,
+				None => return InvalidTransaction::Custom(INVALID_TX_BAD_SUBJECT).into(),
+			};
+
+			match active_vote.has_vote(validator_index) {
+				Some(false) => (),
+				Some(true) => return InvalidTransaction::Custom(INVALID_TX_DOUBLE_VOTE).into(),
+				None => return InvalidTransaction::Custom(INVALID_TX_BAD_VALIDATOR_IDX).into(),
+			}
+
+			ValidTransaction::with_tag_prefix("PvfPreCheckingVote")
+				.priority(T::UnsignedPriority::get())
+				.longevity(
+					TryInto::<u64>::try_into(
+						T::NextSessionRotation::average_session_length() / 2u32.into(),
+					)
+					.unwrap_or(64_u64),
+				)
+				.and_provides((stmt.session_index, stmt.validator_index, stmt.subject))
+				.propagate(true)
+				.build()
+		}
+
+		fn pre_dispatch(call: &Self::Call) -> Result<(), TransactionValidityError> {
+			let stmt = match call {
+				Call::include_pvf_check_statement { stmt, .. } => stmt,
+				_ => return Err(InvalidTransaction::Call.into()),
+			};
+
+			let current_session = shared::Pallet::<T>::session_index();
+			if stmt.session_index < current_session {
+				return Err(InvalidTransaction::Stale.into())
+			} else if stmt.session_index > current_session {
+				return Err(InvalidTransaction::Future.into())
+			}
+
+			let active_vote = match PvfActiveVoteMap::<T>::get(&stmt.subject) {
+				Some(v) => v,
+				None => return Err(InvalidTransaction::Custom(INVALID_TX_BAD_SUBJECT).into()),
+			};
+
+			match active_vote.has_vote(stmt.validator_index.0 as usize) {
+				Some(false) => (),
+				Some(true) => return Err(InvalidTransaction::Custom(INVALID_TX_DOUBLE_VOTE).into()),
+				None => return Err(InvalidTransaction::Custom(INVALID_TX_BAD_VALIDATOR_IDX).into()),
+			}
+
+			Ok(())
+		}
```

Comment:
> Note that here we do not check if the signature is correct also

Yes I have seen this, otherwise you could directly have used the default impl of `pre_dispatch` that just calls `validate`.